torch/_masked/__init__.py

```diff
@@ -512,6 +512,95 @@ def _apply(a):
     return result.coalesce()
 
 
+def _sparse_coo_scatter_reduction_helper(op,
+                                         mask_input: Tensor,
+                                         dims: Tuple[int, ...],
+                                         keepdim: bool,
+                                         dtype: Optional[DType] = None) -> Tensor:
+    reduce = op.__name__
+    assert reduce in {'sum', 'prod', 'mean', 'amax', 'amin'}, \
+        "op must be one of torch.sum, torch.prod, torch.mean, torch.amax or torch.amin"
+
+    output_dtype = dtype
+    values, indices = mask_input._values(), mask_input._indices()
+    input_dims = mask_input.dim()
+    num_sparse_dims = mask_input.sparse_dim()
+    reduced_sparse_dims = []
+    retained_sparse_dims = []
+    reduced_dense_dims = []
+
+    # promote dtype if specified
+    if values.dtype != output_dtype:
+        values = values.to(output_dtype)
+
+    if keepdim:
+        output_shape = tuple((1 if i in dims else mask_input.shape[i]) for i in range(mask_input.ndim))
```

Comment:
This might be easier if you use enumerate
```
if keepdim:
    output_shape = tuple(1 if i in dims else si for (i, si) in enumerate(mask_input.shape))
else:
    output_shape = tuple(si for (i, si) in enumerate(mask_input.shape) if i not in dims)
```